packages/legend-query/src/stores/QueryFunctionsExplorerState.ts

```diff
@@ -0,0 +1,456 @@
+      generateFunctionsExplorerTreeNodeData(
+        queryBuilderState,
+        child,
+        rootPackageName,
+      ),
+    )
+    .forEach((childNode) => {
+      const currentNode = data.nodes.get(childNode.id);
+      if (currentNode) {
+        currentNode.childrenIds = childNode.childrenIds;
+        currentNode.label = childNode.label;
+      } else {
+        data.nodes.set(childNode.id, childNode);
+      }
+    });
+};
+
+export const getFunctionsExplorerTreeData = (
+  root: Package,
+  queryBuilderState: QueryBuilderState,
+  rootPackageName = ROOT_PACKAGE_NAME.MAIN,
+): TreeData<QueryBuilderFunctionsExplorerTreeNodeData> => {
+  const rootIds: string[] = [];
+  const nodes = new Map<string, QueryBuilderFunctionsExplorerTreeNodeData>();
+  const validDisplayablePackageSet = getValidDisplayablePackageSet(
+    queryBuilderState,
+    rootPackageName,
+  );
+  switch (rootPackageName) {
+    case ROOT_PACKAGE_NAME.PROJECT_DEPENDENCY_ROOT:
+      if (
+        queryBuilderState.graphManagerState.graph.dependencyManager.functions
+          .length === 0
+      ) {
+        return { rootIds, nodes };
+      }
+      break;
+    default:
+      if (queryBuilderState.graphManagerState.graph.ownFunctions.length === 0) {
+        return { rootIds, nodes };
+      }
+  }
+
+  root.children
+    .slice()
+    .filter((child) => !(child instanceof Unit))
+    .filter(
+      (child) =>
+        child instanceof Package && validDisplayablePackageSet.has(child),
+    )
+    .sort((a, b) => a.name.localeCompare(b.name))
+    .sort(
+      (a, b) => (b instanceof Package ? 1 : 0) - (a instanceof Package ? 1 : 0),
+    )
+    .forEach((childPackage) => {
+      const childTreeNodeData = generateFunctionsExplorerTreeNodeData(
+        queryBuilderState,
+        childPackage,
+        rootPackageName,
+      );
+      addUniqueEntry(rootIds, childTreeNodeData.id);
+      nodes.set(childTreeNodeData.id, childTreeNodeData);
+    });
+  return { rootIds, nodes };
+};
+
+export const getFunctionsExplorerTreeNodeChildren = (
+  queryBuilderState: QueryBuilderState,
+  node: QueryBuilderFunctionsExplorerTreeNodeData,
+  data: TreeData<QueryBuilderFunctionsExplorerTreeNodeData>,
+  rootPackageName = ROOT_PACKAGE_NAME.MAIN,
+): QueryBuilderFunctionsExplorerTreeNodeData[] => {
+  generateFunctionsExplorerTreeNodeChilrdren(
+    queryBuilderState,
+    node,
+    data,
+    rootPackageName,
+  );
+  return node.childrenIds
+    .map((id) => data.nodes.get(id))
+    .filter(isNonNullable)
+    .sort((a, b) => a.label.localeCompare(b.label))
+    .sort(
+      (a, b) =>
+        (b.packageableElement instanceof Package ? 1 : 0) -
+        (a.packageableElement instanceof Package ? 1 : 0),
+    );
+};
+
+const getAllPackagesFromElement = (element: PackageableElement): Package[] => {
+  if (element.package) {
+    return [element.package].concat(getAllPackagesFromElement(element.package));
+  }
+  return [];
+};
+
+export const generateFunctionSignature = (
```

Comment:
I could see we make use of this in core, let's move it and `generateDefaultValueForPrimitiveParameter` to `DomainHelper`